crates/viewer/re_time_panel/src/data_density_graph.rs

```diff
@@ -442,6 +428,124 @@ pub fn data_density_graph_ui2(
+
+    let mut data = DensityGraphBuilder::new(ui, time_ranges_ui, row_rect);
+
+    // Collect all relevant chunks in the visible time range.
+    // We do this as a separate step so that we can also deduplicate chunks.
+    let visible_time_range = time_ranges_ui
+        .time_range_from_x_range((row_rect.left() - MARGIN_X)..=(row_rect.right() + MARGIN_X));
+
+    // NOTE: These chunks are guaranteed to have data on the current timeline
+    let mut chunk_ranges: Vec<(Arc<Chunk>, ResolvedTimeRange, usize)> = vec![];
+    let mut total_events = 0;
+
+    {
+        visit_relevant_chunks(
+            db,
+            &item.entity_path,
+            item.component_name,
+            timeline,
+            visible_time_range,
+            |chunk, time_range, num_events| {
+                chunk_ranges.push((chunk, time_range, num_events));
+                total_events += num_events;
+            },
+        );
+    }
+
+    // Small chunk heuristics:
+    // We want to render chunks as individual events, but it may be prohibitively expensive
+    // for larger chunks, or if the visible time range contains many chunks.
+    //
+    // We split a large chunk if:
+    // 1. The total number of events is less than some threshold
+    // 2. The number of events in the chunks is less than N, where:
+    //    N is relatively large for sorted chunks
+    //    N is much smaller for unsorted chunks
+
+    {
+        re_tracing::profile_scope!("add_data");
+
+        let can_render_individual_events =
+            config.max_total_chunk_events != 0 && total_events < config.max_total_chunk_events;
+
+        for (chunk, time_range, num_events_in_chunk) in chunk_ranges {
+            let should_render_individual_events = can_render_individual_events
+                && if chunk.is_time_sorted() {
+                    config.max_events_in_sorted_chunk != 0
+                        && num_events_in_chunk < config.max_events_in_sorted_chunk
+                } else {
+                    config.max_events_in_unsorted_chunk != 0
+                        && num_events_in_chunk < config.max_events_in_unsorted_chunk
+                };
+
+            if should_render_individual_events {
+                for (time, num_events) in chunk.num_events_cumulative_per_unique_time(&timeline) {
+                    data.add_chunk_point(time, num_events as usize);
+                }
+            } else {
+                data.add_chunk_range(time_range, num_events_in_chunk);
+            }
+        }
+    }
+
+    data
+}
+
+#[derive(Clone, Copy)]
+pub struct DensityGraphBuilderConfig {
+    pub max_total_chunk_events: usize,
+    pub max_events_in_sorted_chunk: usize,
+    pub max_events_in_unsorted_chunk: usize,
+}
+
+impl DensityGraphBuilderConfig {
+    /// All chunks will be rendered whole.
+    pub const NEVER_SPLIT_CHUNKS: Self = Self {
+        max_total_chunk_events: 0,
+        max_events_in_unsorted_chunk: 0,
+        max_events_in_sorted_chunk: 0,
+    };
```

Comment:
I think something like this is probably a better name

```suggestion
    pub const NEVER_SHOW_INDIVIDUAL_EVENTS: Self = Self {
        max_total_chunk_events: 0,
        max_events_in_unsorted_chunk: 0,
        max_events_in_sorted_chunk: 0,
    };
```